### ISSUE TITLE
Segmentation fault in mdcache_lru_cleanup_try_push (V3.3)

### ISSUE THREAD

**Issue by ghost** [nfs-ganesha/nfs-ganesha#598]
I found a segmentation fault while testing nfs-ganesha V3.3.

The test was mostly the following scripted loop:
1. Create a bunch of zfs filesystems
2. Create a couple of exports (FSAL_VFS) under each using the dbus interface
3. Mount all the exports on a client (NFSv3)
4. Umount the exports
5. Remove the exports using the dbus interface
6. Remove the filesystems

(Some other patterns might have occurred, but the type of filesystem/export/client was the same)

Export ids were generated sequentially and wrapped around after hitting the max. Any existing export ids were skipped. The error occurred sometime after the ids had wrapped around, although certainly not immediately (if relevant).

All export configuration happened in separate files (stored on the temporary zfs filesystems), and the search pattern used when adding the export was the export id.

This is (sadly) all the information I thought to capture when it happened yesterday as I happened to have gdb attached:
```
(gdb) bt
#0  0x00007f70fbee88ed in mdcache_lru_cleanup_try_push (entry=0x7f6fc03df4a0) at /root/sources/nfs-ganesha/src/FSAL/Stackable_FSALs/FSAL_MDCACHE/mdcache_lru.c:1019
#1  0x00007f70fbed4e10 in mdcache_unexport (exp_hdl=0x7f703c2fdb20, root_obj=0x7f701010a098)
    at /root/sources/nfs-ganesha/src/FSAL/Stackable_FSALs/FSAL_MDCACHE/mdcache_export.c:152
#2  0x00007f70fbe456b8 in release_export (export=0x7f703c21da48) at /root/sources/nfs-ganesha/src/support/exports.c:2461
#3  0x00007f70fbe457f9 in unexport (export=0x7f703c21da48) at /root/sources/nfs-ganesha/src/support/exports.c:2488
#4  0x00007f70fbe5e996 in gsh_export_removeexport (args=0x7f7043705340, reply=0x7f703c004840, error=0x7f7043705320)
    at /root/sources/nfs-ganesha/src/support/export_mgr.c:1225
#5  0x00007f70fbef53bd in dbus_message_entrypoint (conn=0x56279f176750, msg=0x7f703c1b77c0, user_data=0x7f70fbf92c10 <export_interfaces>)
    at /root/sources/nfs-ganesha/src/dbus/dbus_server.c:574
#6  0x00007f70fbaa644d in ?? () from /lib/x86_64-linux-gnu/libdbus-1.so.3
#7  0x00007f70fba96b04 in dbus_connection_dispatch () from /lib/x86_64-linux-gnu/libdbus-1.so.3
#8  0x00007f70fba96eb8 in ?? () from /lib/x86_64-linux-gnu/libdbus-1.so.3
#9  0x00007f70fbef610f in gsh_dbus_thread (arg=0x0) at /root/sources/nfs-ganesha/src/dbus/dbus_server.c:809
#10 0x00007f70fbd43609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#11 0x00007f70fbc6a103 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

(gdb) list
1014
1015			if (glist_empty(&entry->export_list)) {
1016				/* it worked */
1017				struct lru_q *q = lru_queue_of(entry);
1018
1019				LRU_DQ_SAFE(lru, q);
1020				entry->lru.qid = LRU_ENTRY_CLEANUP;
1021				atomic_set_uint32_t_bits(&entry->lru.flags,
1022							 LRU_CLEANUP);
1023				/* Note: we didn't take a ref here, so the only ref left

(gdb) p lru
$1 = (mdcache_lru_t *) 0x7f6fc03df6c8
(gdb) p q
$2 = (struct lru_q *) 0x0
(gdb) p entry
$3 = (mdcache_entry_t *) 0x7f6fc03df4a0

(gdb) info locals
q = 0x0
lru = 0x7f6fc03df6c8
qlane = 0x7f70fbfa0ec0 <LRU+2240>
latch = {cp = 0x56279f178020}
__func__ = "mdcache_lru_cleanup_try_push"

(gdb) info args
entry = 0x7f6fc03df4a0

(gdb) print *entry
$6 = {attr_lock = {__data = {__readers = 10, __writers = 0, __wrphase_futex = 2, __writers_futex = 1, __pad3 = 0, __pad4 = 0, __cur_writer = 0, __shared = 0,
      __rwelision = 0 '\000', __pad1 = "\000\000\000\000\000\000", __pad2 = 0, __flags = 0},
    __size = "\n\000\000\000\000\000\000\000\002\000\000\000\001", '\000' <repeats 42 times>, __align = 10}, obj_handle = {handles = {next = 0x7f6fbc52cbf8, 
      prev = 0x7f6fbc5a15b8}, fs = 0x7f703c0a8050, fsal = 0x7f70fbf93080 <MDCACHE>, obj_ops = 0x7f70fbf931e0 <MDCACHE+352>, obj_lock = {__data = {__readers = 0, 
        __writers = 0, __wrphase_futex = 0, __writers_futex = 0, __pad3 = 0, __pad4 = 0, __cur_writer = 0, __shared = 0, __rwelision = 0 '\000', 
        __pad1 = "\000\000\000\000\000\000", __pad2 = 0, __flags = 0}, __size = '\000' <repeats 55 times>, __align = 0}, type = DIRECTORY, fsid = {major = 805952136, 
      minor = 12880284}, fileid = 5, state_hdl = 0x7f6fc03df768}, sub_handle = 0x0, attrs = {request_mask = 9223372036864335822, valid_mask = 1171406, 
    supported = 9560014, type = DIRECTORY, filesize = 3, fsid = {major = 0, minor = 98}, fsid3 = 55320399349144200, acl = 0x0, fileid = 5, mode = 493, numlinks = 3, 
    owner = 33, group = 33, rawdev = {major = 0, minor = 0}, atime = {tv_sec = 1591104250, tv_nsec = 863291242}, creation = {tv_sec = 0, tv_nsec = 0}, ctime = {
      tv_sec = 1591101146, tv_nsec = 850537162}, mtime = {tv_sec = 1583319212, tv_nsec = 300113886}, spaceused = 512, change = 1591101146850537162, generation = 0, 
    expire_time_attr = 60, fs_locations = 0x0, sec_label = {slai_lfs = {lfs_lfs = 0, lfs_pi = 0}, slai_data = {slai_data_len = 0, slai_data_val = 0x0}}}, 
  attr_generation = 0, fh_hk = {node_k = {left = 0x0, right = 0x0, parent = 140116869716266}, key = {hk = 4051788321832339657, fsal = 0x7f70f47cf840 <VFS>, kv = {
        addr = 0x7f6fc028a640, len = 22}}, inavl = false}, mde_flags = 5, attr_time = 1591801982, acl_time = 0, fs_locations_time = 0, lru = {q = {next = 0x0, 
      prev = 0x0}, qid = LRU_ENTRY_NONE, refcnt = 2, flags = 0, lane = 10, cf = 0}, export_list = {next = 0x7f6fc03df6f0, prev = 0x7f6fc03df6f0}, first_export_id = -1, 
  content_lock = {__data = {__readers = 1, __writers = 0, __wrphase_futex = 1, __writers_futex = 0, __pad3 = 0, __pad4 = 0, __cur_writer = 0, __shared = 0, 
      __rwelision = 0 '\000', __pad1 = "\000\000\000\000\000\000", __pad2 = 0, __flags = 0}, __size = "\001\000\000\000\000\000\000\000\001", '\000' <repeats 46 times>, 
    __align = 1}, fsobj = {hdl = {state_lock = {__data = {__readers = 3225286464, __writers = 32623, __wrphase_futex = 3225286464, __writers_futex = 32623,
          __pad3 = 3225286480, __pad4 = 32623, __cur_writer = -1069680816, __shared = 32623, __rwelision = 1 '\001', __pad1 = "\000\000\000\000\000\000", __pad2 = 0, 
          __flags = 0}, __size = "@\367=\300o\177\000\000@\367=\300o\177\000\000P\367=\300o\177\000\000P\367=\300o\177\000\000\001", '\000' <repeats 22 times>, 
        __align = 140117943383872}, no_cleanup = false, {file = {obj = 0x0, list_of_states = {next = 0x0, prev = 0x0}, layoutrecall_list = {next = 0x0, prev = 0x0}, 
          lock_list = {next = 0x0, prev = 0x7f6fc03df7b0}, nlm_share_list = {next = 0x7f6fc03df7b0, prev = 0x0}, write_delegated = false, write_deleg_client = 0x0, 
          fdeleg_stats = {fds_curr_delegations = 0, fds_deleg_type = OPEN_DELEGATE_NONE, fds_delegation_count = 0, fds_recall_count = 0, fds_avg_hold = 0, 
            fds_last_delegation = 0, fds_last_recall = 0, fds_num_opens = 0, fds_first_open = 0}, cbgetattr = {state = CB_GETATTR_NONE, change = 0, filesize = 0, 
            modified = false}, anon_ops = 0}, dir = {junction_export = 0x0, export_roots = {next = 0x0, prev = 0x0}, exp_root_refcount = 0}}}, fsdir = {chunks = {
        next = 0x7f6fc03df740, prev = 0x7f6fc03df740}, detached = {next = 0x7f6fc03df750, prev = 0x7f6fc03df750}, spin = 1, detached_count = 0, dhdl = {state_lock = {
          __data = {__readers = 0, __writers = 0, __wrphase_futex = 0, __writers_futex = 0, __pad3 = 0, __pad4 = 0, __cur_writer = 0, __shared = 0, 
            __rwelision = 0 '\000', __pad1 = "\000\000\000\000\000\000", __pad2 = 0, __flags = 0}, __size = '\000' <repeats 55 times>, __align = 0}, no_cleanup = false, {
          file = {obj = 0x0, list_of_states = {next = 0x7f6fc03df7b0, prev = 0x7f6fc03df7b0}, layoutrecall_list = {next = 0x0, prev = 0x0}, lock_list = {next = 0x0, 
              prev = 0x0}, nlm_share_list = {next = 0x0, prev = 0x0}, write_delegated = false, write_deleg_client = 0x0, fdeleg_stats = {fds_curr_delegations = 0,
              fds_deleg_type = OPEN_DELEGATE_NONE, fds_delegation_count = 0, fds_recall_count = 0, fds_avg_hold = 0, fds_last_delegation = 0, fds_last_recall = 0, 
              fds_num_opens = 0, fds_first_open = 0}, cbgetattr = {state = CB_GETATTR_NONE, change = 0, filesize = 0, modified = false}, anon_ops = 0}, dir = {
            junction_export = 0x0, export_roots = {next = 0x7f6fc03df7b0, prev = 0x7f6fc03df7b0}, exp_root_refcount = 0}}}, parent = {addr = 0x7f6f70041cd0, len = 22}, 
      parent_time = 0, first_ck = 0, avl = {t = {root = 0x0, cmp_fn = 0x7f70fbeeec71 <avl_dirent_name_cmpf>, height = -1, first = 0x0, last = 0x0, size = 0}, ck = {
          root = 0x0, cmp_fn = 0x7f70fbeeed08 <avl_dirent_ck_cmpf>, height = -1, first = 0x0, last = 0x0, size = 0}, sorted = {root = 0x0, 
          cmp_fn = 0x7f70fbeeed7f <avl_dirent_sorted_cmpf>, height = -1, first = 0x0, last = 0x0, size = 0}, collisions = 0}}}}
```

**Comment by dang** [comment]
It looks like an entry got reaped while it was in the export map.  Not sure how this could happen.  I don't suppose you have logs?

**Comment by ghost** [comment]
I do have some logs still, the last few lines before the crash are:
```
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[svc_188] vfs_check_handle :FSAL :DEBUG :Found filesystem /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4f7WAMK8M for handle for FSAL VFS
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] process_claim :FSAL :INFO :FSAL VFS Claiming /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4gob3tk2Z
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] claim_posix_filesystems :FSAL :INFO :Root fs for export /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4gob3tk2Z/nfs is /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4gob3tk2Z
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] vfs_lookup_path :FSAL :DEBUG :filesystem /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4gob3tk2Z for path /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4gob3tk2Z/nfs
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] fsal_common_is_referral :FSAL :DEBUG :Checking attrs for referral, handle: 0x7f703c2ff5f0, valid_mask: 11dfce, request_mask: 91dfce, supported: 91dfce
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] free_vfs_fsal_obj_handle :FSAL :DEBUG :Releasing obj_hdl=0x7f6fd83e9e60, myself=0x7f6fd83e9e60
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[svc_160] vfs_check_handle :FSAL :DEBUG :Found filesystem /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4f7WAMK8M for handle for FSAL VFS
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] free_vfs_fsal_obj_handle :FSAL :DEBUG :Releasing obj_hdl=0x7f703c20b220, myself=0x7f703c20b220
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] populate_posix_file_systems :FSAL :DEBUG :File systems are initialized
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] process_claim :FSAL :DEBUG :FSAL VFS Repeat Claiming /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4gob3tk2Z
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] claim_posix_filesystems :FSAL :INFO :Root fs for export /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4gob3tk2Z/fop is /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4gob3tk2Z
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] resolve_posix_filesystem :FSAL :DEBUG :Not trying to claim filesystems again because retval != EAGAIN Success(0)
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] vfs_lookup_path :FSAL :DEBUG :filesystem /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4gob3tk2Z for path /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4gob3tk2Z/fop
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] fsal_common_is_referral :FSAL :DEBUG :Checking attrs for referral, handle: 0x7f703c14abf0, valid_mask: 11dfce, request_mask: 91dfce, supported: 91dfce
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] free_vfs_fsal_obj_handle :FSAL :DEBUG :Releasing obj_hdl=0x7f70d005fe30, myself=0x7f70d005fe30
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[dbus_heartbeat] free_vfs_fsal_obj_handle :FSAL :DEBUG :Releasing obj_hdl=0x7f703c114c00, myself=0x7f703c114c00
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[svc_192] vfs_check_handle :FSAL :DEBUG :Found filesystem /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4f7WAMK8M for handle for FSAL VFS
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[svc_150] vfs_check_handle :FSAL :DEBUG :Found filesystem /export/testnfsd/clone_f03d8314c8251da9622a7268996e032f7b86313c_XuD4f7WAMK8M for handle for FSAL VFS
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[svc_150] fsal_common_is_referral :FSAL :DEBUG :Checking attrs for referral, handle: 0x7f70e8462990, valid_mask: 11dfce, request_mask: 800000000091dfce, supported: 91dfce
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[svc_150] free_vfs_fsal_obj_handle :FSAL :DEBUG :Releasing obj_hdl=0x7f6f7000b350, myself=0x7f6f7000b350
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[svc_150] fsal_common_is_referral :FSAL :DEBUG :Checking attrs for referral, handle: 0x7f70e844bef0, valid_mask: 11dfce, request_mask: 800000000091dfce, supported: 91dfce
10/06/2020 17:13:06 : epoch 5ee0c175 : devnfs2 : ganesha.nfsd-190830[svc_150] free_vfs_fsal_obj_handle :FSAL :DEBUG :Releasing obj_hdl=0x7f6f70377500, myself=0x7f6f70377500

```

**Comment by dang** [comment]
That doesn't seem to include anything on releasing an export, although it does have logs from creating an export.  Frank, any ideas?

**Comment by ffilz** [comment]
I have been working on significant improvements to dynamic export management. Along the way I decided to try and improve the handling op contexts, and through that I have tripped on some export refcounting issues. I am now auditing EVERY export reference and finding lots of bugs.

One of the bugs I have identified is an extraneous call to cih_remove_checked in mdcache_unexport. If your multiple exports ever result in an mdcache node being the root of more than one export, that is definitely a problem. I think it could also be a problem if an mdcache node is a root of one export but also visible and cached via another export.

But your issue could also be caused by any of a number of the other export refcounting issues I've hit. 

The audit task has become larger than I expected but will hopefully be complete next week and then we can look through and see what can be back ported to 3.3. I will try and tease out the refcounting fixes from the re-work of how op contexts are handled so that we can have something to backport for 3.4 though maybe it would be good to backport the op context management, something for discussion when all is done).
